### ISSUE TITLE
Issue with Gangsters Organized Crime and the Experimental version

### ISSUE THREAD

**Issue by elishacloud** [narzoul/DDrawCompat#18]
The latest experimental version has an issue with the game [Gangsters Organized Crime](https://www.gog.com/game/gangsters_organized_crime).  It does not show any of the text on the buttons or the labels next to the buttons.

Here is what the experimental version looks like:
![gangsters-ddrawcompat](https://user-images.githubusercontent.com/25625456/28700882-517433e4-7307-11e7-8dbf-7b2012f31573.png)

Here is what DDrawCompat v0.2.1 looks like:
![gangsters](https://user-images.githubusercontent.com/25625456/28700910-7e6d3b70-7307-11e7-81e7-ba0eb0adda18.png)

**Comment by narzoul** [comment]
I haven't yet looked into the actual issue as I got sidetracked by how this game bypasses loading the wrapper in the first place. I traced down the problem to a windows compatibility shim made specifically for this game, which is implemented by AcSpecfc.dll. The problem is that this dll gets loaded early into the process and it implicilty depends on ddraw.dll. Although it's not on the KnownDLLs list, it still seems to be somehow protected as it can't be replaced by an app-local version either, and it seems the protection extends to its dependencies as those are loaded only from the system directory also.

As a quick and dirty fix for this particular game, of course you can just rename the executable file, but that would disable the builtin shim too (though I didn't notice it causing any issues yet).

I looked extensively into application and assembly manifests as a possible generic solution to problems like this, including the COM hooking issue with WH40k and other cases such as loading ddraw.dll with a full path explicitly from the system directory. I think good results could be reached, especially when combining with my previous idea of using dciman32.dll as a support library for handling the redirection.
I'll try to implement this idea first, probably as a separate project as it could potentially be used with other DirectDraw wrappers too.

While testing out this method of redirection for ddraw.dll, I tried to take a very quick preliminary look into the reported problem too, but the first issue I ran into was actually an apparent reference count leak somewhere in the TagSurface class, which prevented the game from starting up in the first place. I wonder how you didn't run into that one...

**Comment by elishacloud** [comment]
I am using dsound.dll as the wrapper and hooking the ddraw.dll APIs from this dll.  I am hooking the APIs in the `DllMain` function when `DLL_PROCESS_ATTACH` is called.  From what I can tell with this game the `DLL_PROCESS_ATTACH` will away be called in dsound.dll before any of the required ddraw.dll APIs are called.  I am using [hot patching](https://github.com/elishacloud/dxwrapper/blob/master/Hooking/HotPatch.cpp) and fail over to [IAT patching](https://github.com/elishacloud/dxwrapper/blob/master/Hooking/IATPatch.cpp) if that fails.  This seems to work fine for me.  No idea why I am not seeing the reference counter issue.  So far I have never had an issue with the game failing to load.  I did write up with some details on it [here](https://github.com/elishacloud/dxwrapper/wiki/Gangsters-Organized-Crime).

**Comment by elishacloud** [comment]
Now it seems that with the latest experimental version Gangsters Organized Crime will not run at all.  It works fine with both v0.2.1 and v0.2.0, but gives me an error message with the latest experimental version.

I am getting this error:

![error](https://user-images.githubusercontent.com/25625456/37560426-0b058336-29f5-11e8-9433-98b90ba8e228.png)

**Comment by narzoul** [comment]
Latest experimental version works mostly fine for me now. The scroll bars in the options menu are still not updating and the mouse seems rather sluggish in-game (not in the menus), but otherwise the game seems playable.

I've simply renamed the executable for testing so that the game loads DDrawCompat and not the system ddraw.dll. I did try to make a dciman32.dll proxy about a year ago and experimented with the manifest stuff, but I unfortunately lost it to a hard drive malfunction. I will try to look into these remaining issues again later.
